tests/expose_test.go

```diff
@@ -536,47 +536,99 @@ var _ = Describe("[rfe_id:253][crit:medium][vendor:cnv-qe@redhat.com][level:comp
-			err = virtctl()
-			Expect(err).ToNot(HaveOccurred())
+			virtctl := tests.NewRepeatableVirtctlCommand("start", "--namespace", namespace, name)
+			Expect(virtctl()).To(Succeed(), "should succeed starting a VMI via `virtctl start ...`")
 
 			By("Getting the status of the VMI")
 			Eventually(func() bool {
-				vm, err = virtClient.VirtualMachine(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vm, err = virtClient.VirtualMachine(namespace).Get(name, &k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return vm.Status.Ready
 			}, 120*time.Second, 1*time.Second).Should(BeTrue())
 
 			By("Getting the running VMI")
 			var vmi *v1.VirtualMachineInstance
 			Eventually(func() bool {
-				vmi, err = virtClient.VirtualMachineInstance(vm.Namespace).Get(vm.Name, &k8smetav1.GetOptions{})
+				vmi, err = virtClient.VirtualMachineInstance(namespace).Get(name, &k8smetav1.GetOptions{})
 				Expect(err).ToNot(HaveOccurred())
 				return vmi.Status.Phase == v1.Running
 			}, 120*time.Second, 1*time.Second).Should(BeTrue())
 
-			tests.GenerateHelloWorldServer(vmi, testPort, "tcp")
-		})
+			return vmi
+		}
+
+		createStoppedVM := func(virtClient kubecli.KubevirtClient, namespace string) error {
+			By("Creating an VM object")
+			template := newLabeledVMI("vm", virtClient, false)
+			vm = tests.NewRandomVirtualMachine(template, false)
+
+			By("Creating the VM")
+			vm, err = virtClient.VirtualMachine(namespace).Create(vm)
+			return err
+		}
+
+		startVMWithServer := func(protocol string, port int) *v1.VirtualMachineInstance {
+			By("Calling the start command on the stopped VM")
+			vmi := startVMIFromVMTemplate(vm.GetName(), vm.GetNamespace())
+			if vmi == nil {
+				return nil
+			}
+			tests.GenerateHelloWorldServer(vmi, port, protocol)
+			return vmi
+		}
 
 		Context("Expose a VM as a ClusterIP service.", func() {
-			It("[test_id:1538][label:masquerade_binding_connectivity]Connect to ClusterIP service that was set when VM was offline.", func() {
+			tests.BeforeAll(func() {
+				tests.BeforeTestCleanup()
+			})
+
+			BeforeEach(func() {
+				Expect(createStoppedVM(virtClient, tests.NamespaceTestDefault)).To(Succeed(), "should create a stopped VM.")
+			})
+
+			BeforeEach(func() {
+				vmExposeArgs = []string{
+					expose.COMMAND_EXPOSE,
+					"virtualmachine", "--namespace", vm.GetNamespace(), vm.GetName(),
+					"--port", servicePort, "--name", serviceName, "--target-port", strconv.Itoa(testPort),
+				}
+			})
+
+			AfterEach(func() {
+				Expect(jobs).NotTo(BeEmpty(), "a successful test must have stored a way to delete the batchv1.Job entity")
+				Expect(cleanupJobs(jobs)).To(BeEmpty(), "should be able to delete the multiple k8sv1.`Job` entities")
+			})
+
+			AfterEach(func() {
+				Expect(cleanupService(serviceName, vm.GetNamespace())).To(Succeed(), "should be able to delete the k8sv1.Service entity")
+			})
+
+			AfterEach(func() {
+				if vm != nil {
```

Comment:
I've tried an approach relying on calling `BeforeTestCleanup` and it bloats the test execution time by > 10 minutes.

I would avoid it for that reason.

Reply:
Why? You don't have any shared resources between the tests.

Reply:
I'm guessing creating storage for the VMI and all those things take time. And doing it once per test context is a lot cheaper than doing it twice per test.

Reply:
I guess I miss something. But here in this specific context I don't see you are creating share storage for the context. The BeforeAll is only calling the cleanup. Isn't it?

Reply:
I mistakenly assumed you meant this approach on the entire file, not limited to the VM tests. 

You are right; I will take the `BeforeTestsCleanup` route.

Reply:
I have to agree here with Alona, we cannot use the teardown cleanup approach in general per the info collection on failure.
I do not know which is worse really... not having logs or leaking resources.

Reply:
So let's agree, because the current implementation leaks resources and logs more than it should (for instance, the 3 VMI tests are executed sequentially without tearing down any resources, since the `BeforeTestsCleanup` is executed **only** on the `tests.BeforeAll` clauses ). 

Any alternative is better than what we have right now (e.g. if it currently logs too much and leaks resources, just leaking resources is better).